dra-evolution/pkg/api/capacity_types.go

```diff
@@ -3,168 +3,110 @@ package api
 import (
 	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/types"
 )
 
-const (
-	DevMgmtAPIVersion = "devmgmtproto.k8s.io/v1alpha1"
-)
-
-// DevicePool represents a collection of devices managed by a given driver. How
+// ResourcePool represents a collection of devices managed by a given driver. How
 // devices are divided into pools is driver-specific, but typically the
 // expectation would a be a pool per identical collection of devices, per node.
 // It is fine to have more than one pool for a given node, for the same driver.
-type DevicePool struct {
-	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
+type ResourcePool struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 
-	Spec   DevicePoolSpec   `json:"spec,omitempty"`
-	Status DevicePoolStatus `json:"status,omitempty"`
+	Spec ResourcePoolSpec `json:"spec"`
+
+	// Stretch goal for 1.31: define status
 }
 
-// DevicePoolSpec identifies the driver and contains the data for the pool
-// prior to any allocations.
-// NOTE: It's not clear that spec/status is the right model for this data.
-type DevicePoolSpec struct {
-	// NodeName is the name of the node containing the devices in the pool.
-	// For network attached devices, this may be empty.
-	// +optional
-	NodeName *string `json:"nodeName,omitempty"`
+type ResourcePoolSpec struct {
+	ResourcePoolAccessability `json:",inline"`
```

Comment:
The nice thing about keeping embedded structs is that you can add the following to such structs to _enforce_ their one-of semantics.
```
// +kubebuilder:validation:MaxProperties=1
```
For example ...
```
// DeviceAttribute is a combination of an attribute name and its value.
type DeviceAttribute struct {
	Name string `json:"name"
	DeviceAttributeValue `json:",inline"
}

// DeviceAttributeValue must have one and only one field set.
// +kubebuilder:validation:MaxProperties=1
type DeviceAttributeValue struct {
	QuantityValue *resource.Quantity `json:"quantity,omitempty"
	BoolValue *bool `json:"bool,omitempty"
	StringValue *string `json:"string,omitempty"
	VersionValue *string `json:"version,omitempty"
}
```

The following looks nicer in go, but you lose the ability to force users to only set one of the value types.
```
// DeviceAttribute is a combination of an attribute name and its value.
type DeviceAttribute struct {
	Name string `json:"name"

	QuantityValue *resource.Quantity `json:"quantity,omitempty"
	BoolValue *bool `json:"bool,omitempty"
	StringValue *string `json:"string,omitempty"
	VersionValue *string `json:"version,omitempty"
}
```

Reply:
I'll remove one-ofs where we currently only have one field (`ResourcePoolAccessability`), but keep the struct where we clearly have many (`DeviceAttributeValue`).

Reply:
For now I'm keeping the struct when they appear inside a slice, but we can iterate on this further.

Reply:
I think we should keep them everywhere we *know* we will have alternate fields in the future (even if for now there is only 1). Otherwise we lose the ability to decorate such structs with `// +kubebuilder:validation:MaxProperties=1
` to enforce their "one-of-ness".

Reply:
See https://github.com/kubernetes-sigs/wg-device-management/pull/14#discussion_r1623902425.

Reply:
`+kubebuilder` tags should not be used.

I also avoided adding `+oneof` because it hasn't been formalized.